pkg/storage/stores/shipper/gateway_client.go

```diff
@@ -136,6 +170,56 @@ func (s *GatewayClient) doQueries(ctx context.Context, queries []chunk.IndexQuer
 	return nil
 }
 
+// ringModeDoQueries prepares an index query to be sent to the Index Gateway, and then sends it
+// using the clientDoQueries implementation.
+//
+// The preparation and sending phase includes:
+// 1. Extracting the tenant name from the query.
+// 2. Fetching different Index Gateway instances assigned to the extracted tenant.
+// 3. Iterating in parallel over all fetched Index Gateway instances, getting their gRPC connections
+//  from the pool and invoking clientDoQueries using their client.
+func (s *GatewayClient) ringModeDoQueries(ctx context.Context, gatewayQueries []*indexgatewaypb.IndexQuery, queryKeyQueryMap map[string]chunk.IndexQuery, callback chunk.QueryPagesCallback) error {
+	// TODO: should we keep iterating if an error occur?
+	for _, q := range gatewayQueries {
+		bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+
+		splittedHashed := strings.Split(q.HashValue, ":")
+		if len(splittedHashed) < 2 {
+			level.Error(util_log.Logger).Log("msg", "query hash value in incorrect format", "hash_value", q.HashValue)
+			continue
+		}
+
+		// tenant is the second parameter.
+		tenantID := splittedHashed[1]
+		key := util.TokenFor(tenantID, "" /* labels */)
+		rs, err := s.ring.Get(key, ring.WriteNoExtend, bufDescs, bufHosts, bufZones)
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", "failed to get index key", "error", err)
+			continue
+		}
+
+		_, err = rs.Do(ctx, 0 /* delay, TODO: what is it for? */, func(ctx context.Context, instanceDesc *ring.InstanceDesc) (interface{}, error) {
```

Comment:
`rs.Do` would query all the index gateways which is not required. We just want one successful response from one of the index gateway from the replica set.
We should try just one and if it fails try the next one until we have tried all of them.
We should do it with some randomization to avoid querying the same index gateway from a replica set for all the queries.

Reply:
thanks for the comment!

double checking: are you sure we don't want to execute for all index gateway instances in parallel? Because this way it looks like a hedging request, which I believe is something we want.

Ex: Imagine we have 20 index gateway instances, and instances gateways 4, 5 and 6 are assigned to tenant ABC. Right now we have this problem of having specific index gateway instances taking really long to answer a query (1min+). If we sent the request in parallel to the three gateways (4, 5 and 6), we overcome this issue as long as the three instances aren't problematic at the same time. If we first sent to one index gateway (let's say Index Gateway 4), and only then if it fails we proceed to send to the next one (let's say Index Gateway 5), we will still have the same problem.

Does it make sense? :thinking:

Reply:
Hedging works by redoing only the queries which have not received a response within a specific duration. This could DOS index gateway by 3Xing the QPS(assuming RF=3).

Reply:
Good call! Changed it to not send all requests all the time, WDYT?